### ISSUE TITLE
Application error on startup( 1.40).

### ISSUE THREAD

**Issue by phyrz91** [chenxiaolong/BCR#303]
Application error on startup. 1.39 works well.  Android 13. Xiaomi 12
```
```
type: crash
osVersion: Xiaomi/kenvyra_cupid/cupid:13/TQ2A.230405.003.E1/eng.jens.20230418.002250:user/release-keys
package: com.chiller3.bcr:75520
process: com.chiller3.bcr
processUptime: 86 + 49 ms

java.lang.RuntimeException: Unable to start activity ComponentInfo{com.chiller3.bcr/com.chiller3.bcr.SettingsActivity}: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3652)
	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3789)
	at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:101)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:138)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2305)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:201)
	at android.os.Looper.loop(Looper.java:288)
	at android.app.ActivityThread.main(ActivityThread.java:7947)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:968)
Caused by: java.lang.IllegalStateException: You need to use a Theme.AppCompat theme (or descendant) with this activity.
	at androidx.appcompat.app.AppCompatDelegateImpl.ensureSubDecor(Unknown Source:676)
	at androidx.appcompat.app.AppCompatDelegateImpl.setContentView(SourceFile:6)
	at androidx.appcompat.app.AppCompatActivity.setContentView(SourceFile:2)
	at com.chiller3.bcr.SettingsActivity.onCreate(Unknown Source:6)
	at android.app.Activity.performCreate(Activity.java:8388)
	at android.app.Activity.performCreate(Activity.java:8367)
	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1425)
	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3633)
	... 12 more

```
```

**Comment by chenxiaolong** [comment]
Thos was reported for an older version in #275 too. I have no idea what might be causing this or how to fix it. It seems to only happen on MIUI. It's complaining that BCR's theme doesn't use `Theme.AppCompat`, but it does (ever since the very first version 1.0).

**Comment by phyrz91** [comment]
Hi, thx for reply. It's not MiUI, Kenvyra(fork LOS 20).

**Comment by ha33y3** [comment]
I've also had the same problem
It crush before it opened the app
Android 13

**Comment by Jollfye** [comment]
Looks like the same error, going back to 39. Here are some logs.
[crash.log.txt](https://github.com/chenxiaolong/BCR/files/11290661/crash.log.txt)
[Log_2023-04-21_03-26-08.txt](https://github.com/chenxiaolong/BCR/files/11290664/Log_2023-04-21_03-26-08.txt)
![Screenshot_20230421-033502_Settings](https://user-images.githubusercontent.com/6180818/233493668-e721302c-3c9b-490b-8ec4-b170acb12e53.png)

**Comment by dhrubo66** [comment]
+1
Same issue here as well on v1.40. App crushes at startup. And also app name replaced with package name. Using Android 13 based PixelOS ROM. Downgraded to v1.39 and it works without any issue.

**Comment by chenxiaolong** [comment]
Hmm, maybe not. I still can't reproduce this when setting my device to each of the languages that BCR supports.

**Comment by chenxiaolong** [comment]
I just finished testing 1.39 -> 1.40 upgrades and 1.40 clean installs while the phone is set to each of the 9 languages BCR supports on my Pixel 7 Pro. I can't reproduce the crash at all. Same thing in the Android emulator.

I have absolutely no idea what would be causing this. Are there any Magisk modules or theming-related settings in the ROM that might be interfering?

**Comment by dhrubo66** [comment]
Here are a few notes from me. I don't know if they'd be useful to you:
- I have no theme related module, however I disabled all the modules, did an update + clean install. The issue was there in both the cases
- Application name is "com.chiller3.bcr.RecorderApplication" instead of "BCR"
- Application version in app info shows 1.39 where it was suppose to show 1.40 (both in update and clean install)
- I have attached a photo of logs, apologies because I've never taken a logcat before and don't know if I did it right.

**Comment by Jollfye** [comment]
Modules I have installed, no theming, clean PE+ and latest magisk 26.1.
![Screenshot_20230422-021028_Manager](https://user-images.githubusercontent.com/6180818/233726693-fa91fee7-b4f8-457d-a5c8-93f0dd7bfe09.png)
![Screenshot_20230422-021033_Manager](https://user-images.githubusercontent.com/6180818/233726696-ef27b1e8-3a33-4a34-9c1f-aeb7a6b4ddf3.png)

**Comment by chenxiaolong** [comment]
@dhrubo66

> Application version in app info shows 1.39 where it was suppose to show 1.40 (both in update and clean install)

That is very interesting and seems like it might potentially be the problem. If you're familiar with `adb shell` or have Termux installed, can you run this:

```bash
# For adb shell:
pm dump com.chiller3.bcr > /sdcard/bcr_dump.txt
# For termux:
su -c 'pm dump com.chiller3.bcr' > /sdcard/bcr_dump.txt
```

and then upload `/sdcard/bcr_dump.txt`?

**Comment by phyrz91** [comment]
Latest Magisk, v 1.41.
[bcr_dump.txt](https://github.com/chenxiaolong/BCR/files/11300213/bcr_dump.txt)
[crash.log (12).txt](https://github.com/chenxiaolong/BCR/files/11300214/crash.log.12.txt)
[bcr1.41.txt](https://github.com/chenxiaolong/BCR/files/11300215/bcr1.41.txt)
![Screenshot_20230422-055541_Kenvyra Launcher](https://user-images.githubusercontent.com/25689133/233761002-4a9934f4-210a-44d8-bae1-8ed743b532f8.png)

**Comment by Rtizer-9** [comment]
@chenxiaolong

Temporary SOLUTION - Just install the 1.41 apk over 1.39 and everything works as it's supposed to.

I'm facing the exact same error and to me the most surprising thing is that if zip only contains 1.41 apk how come installed app shows version 1.39. Is there any other file which also defines what version is gonna be shown in app settings page?

**Comment by Rtizer-9** [comment]
@chenxiaolong is this helpful? Lots of answers here - https://stackoverflow.com/questions/39604889/how-to-fix-you-need-to-use-a-theme-appcompat-theme-or-descendant-with-this-a

**Comment by symbuzzer** [comment]
Same issue on 1.41 :(

**Comment by chenxiaolong** [comment]
@phyrz91 Thanks for the new logs! Unfortunately, it looks like the `pm dump` command failed to run. If you used termux to run the command, can you try adb instead? (or vice versa)

---

> Temporary SOLUTION - Just install the 1.41 apk over 1.39 after installing 1.41 update zip and everything works as it's supposed to.
>
> I'm facing the exact same error and to me the most surprising thing is that if zip only contains 1.41 apk how come installed app shows version 1.39. Is there any other file which also defines what version is gonna be shown in app settings page?

That's interesting. There must be corrupted state left somewhere on the device. While it's still in the broken state, can you use a root file manager to grab `/data/system/packages.xml` and upload that? It does not contain any private information, but does contain a list of all installed apps and their state (eg. the permissions granted to them). If you prefer not to upload it publicly, feel free to email to `accounts+github AT chiller3.com` instead.

(If anyone else is willing to upload/send their copy of the file, that'd be helpful too--More things to compare.)

**Comment by Rtizer-9** [comment]
I've actually done what I said above - I've installed 1.41 apk when after installing 1.41 zip it wasn't working so now everything works. Is this the broken state you're referring to?

**Comment by Rtizer-9** [comment]
Also did you look at the stack overflow link I posted. I checked the androidmanifest.xml file and maybe some answers posted in the link may apply to that but you know better.

**Comment by dhrubo66** [comment]
@chenxiaolong 
Here are my logcats from both termux and adb-shell. But they seem to be similar to @phyrz91 
I have also e-mailed my copy of packages.xml
[bcr_dump-(adb-shell).txt](https://github.com/chenxiaolong/BCR/files/11300419/bcr_dump-.adb-shell.txt)
[bcr_dump-(termux).txt](https://github.com/chenxiaolong/BCR/files/11300420/bcr_dump-.termux.txt)

**Comment by phyrz91** [comment]
> I've actually done what I said above - I've installed 1.41 apk when after installing 1.41 zip it wasn't working so now everything works. Is this the broken state you're referring to?
> 
> A possible corrupt state (files from previous version) might be the magisk module directory but it gets removed normally when you uninstall the module from magisk manager. I even clean installed the 1.41 zip but the error persists.

I confirm, after installing the zip module, I installed the apk extracted from the zip, everything works.

 It works even after reinstalling the zip through Magisk.

**Comment by chenxiaolong** [comment]
> A possible corrupt state (files from previous version) might be the magisk module directory but it gets removed normally when you uninstall the module from magisk manager. I even clean installed the 1.41 zip but the error persists.

Yep, during uninstallation, the magisk module directory is completely removed. After rebooting, Android's is supposed to wipe the data for an app that no longer exists and remove it from the package manager state (`/data/system/packages.xml`). But I've seen situations where that has not happened before. 

> Also did you look at the stack overflow link I posted. I checked the androidmanifest.xml file and maybe some answers posted in the link may apply to that but you know better.

I did. BCR's theme is Google's Material 3 with almost no modifications:

https://github.com/chenxiaolong/BCR/blob/2c752133d1c1a882c25260d0169797fbfce8fd04/app/src/main/res/values/themes.xml#L2

and the Material 3 theme is directly derived from `Theme.AppCompat`. I don't think this is an actual code issue, just that old resources from 1.39 are somehow being used with 1.40/1.41.

**Comment by chenxiaolong** [comment]
For folks on 1.41 who:

* did not do the workaround of manually installing the apk (so BCR is still broken)
* OR did not reboot yet after doing the workaround

Can you please upload `/data/local/tmp/bcr_remove_hard_restrictions.log`?

**Comment by chenxiaolong** [comment]
Please give this test build a try: [BCR-1.41.r6.g6357df4-release.zip](https://github.com/chenxiaolong/BCR/files/11302524/BCR-1.41.r6.g6357df4-release.zip)

1. First, make sure BCR is still crashing. If you did any workarounds to get things working, please undo them.
2. Flash the test build.
3. Reboot.
4. If BCR still crashes, please upload both log files beginning with `bcr_` in `/data/local/tmp`.

**Comment by phyrz91** [comment]
> For folks on 1.41 who:
> 
>     * did not do the workaround of manually installing the apk (so BCR is still broken)
> 
>     * OR did not reboot yet after doing the workaround
> 
> 
> Can you please upload `/data/local/tmp/bcr_remove_hard_restrictions.log`?

[bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11302905/bcr_remove_hard_restrictions.log)

Test build works well.

**Comment by chenxiaolong** [comment]
Thanks for uploading that log, @phyrz91!

That seems to confirm that this is an obscure bug in Android. It saw the new APK:

```
    codePath=/system/priv-app/com.chiller3.bcr
...
      firstInstallTime=2023-04-23 07:32:35
```

but still had information about the old version cached:

```
    versionCode=75520 minSdk=28 targetSdk=33
...
    versionName=1.39
```

The test build tries to work around this by forcibly deleting the package manager cache for BCR on every boot. If the test build works well for other folks too, I'll merge those changes and release 1.42.

**Comment by Jollfye** [comment]
Was on 39, installed 41 and verified the issue is present, installed the test build, restarted system a couple of times just for sure, issue is gone, all works fine 🎉
[bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11303039/bcr_remove_hard_restrictions.log)
[bcr_clear_package_manager_caches.log](https://github.com/chenxiaolong/BCR/files/11303041/bcr_clear_package_manager_caches.log)

**Comment by chenxiaolong** [comment]
> Was on 39, installed 41 and verified the issue is present, installed the test build, restarted system a couple of times just for sure, issue is gone, all works fine 🎉
> [bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11303039/bcr_remove_hard_restrictions.log)
> [bcr_clear_package_manager_caches.log](https://github.com/chenxiaolong/BCR/files/11303041/bcr_clear_package_manager_caches.log)

Thanks for testing! Once a couple more people test and post their logs, I'll know for sure, but I have a strong suspicion this is the root cause:

(from `bcr_clear_package_manager_caches.log`)
```
Timestamp: Fri Aug  6 02:38:46 +06 1971
```

I'm guessing the ROM does not set the time correctly until too late in the boot process, which causes Android to not properly update the package manager cache for BCR's apk.

**Comment by ag6ag1** [comment]
> > Was on 39, installed 41 and verified the issue is present, installed the test build, restarted system a couple of times just for sure, issue is gone, all works fine 🎉
> > [bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11303039/bcr_remove_hard_restrictions.log)
> > [bcr_clear_package_manager_caches.log](https://github.com/chenxiaolong/BCR/files/11303041/bcr_clear_package_manager_caches.log)
> 
> Thanks for testing! Once a couple more people test and post their logs, I'll know for sure, but I have a strong suspicion this is the root cause:
> 
> (from `bcr_clear_package_manager_caches.log`)
> 
> ```
> Timestamp: Fri Aug  6 02:38:46 +06 1971
> ```
> 
> I'm guessing the ROM does not set the time correctly until too late in the boot process, which causes Android to not properly update the package manager cache for BCR's apk.

seems the same problem of vanced manager when the YouTube app got un-installed on every boot before adding the settings in manager itself...

**Comment by chenxiaolong** [comment]
Would a couple folks mind testing this build? [BCR-1.41.r16.g897e5fb-release.zip](https://github.com/chenxiaolong/BCR/files/11376910/BCR-1.41.r16.g897e5fb-release.zip) This build reduces the amount of hacky stuff BCR does to work around this issue, but hopefully it still works.

Please report if BCR is working or not with this build and also upload the 2 BCR logs from `/data/local/tmp/` (even if BCR is working).

Thanks!

**Comment by dhrubo66** [comment]
> Would a couple folks mind testing this build? [BCR-1.41.r16.g897e5fb-release.zip](https://github.com/chenxiaolong/BCR/files/11376910/BCR-1.41.r16.g897e5fb-release.zip) This build reduces the amount of hacky stuff BCR does to work around this issue, but hopefully it still works.
> 
> Please report if BCR is working or not with this build and also upload the 2 BCR logs from `/data/local/tmp/` (even if BCR is working).
> 
> Thanks!
> 
> EDIT: No need to do anything fancy (eg. downgrade, clean flash, etc). Just install as normal.

Did a dirty flash. Working normally. Here are the logs:
[bcr_clear_package_manager_caches.log](https://github.com/chenxiaolong/BCR/files/11377102/bcr_clear_package_manager_caches.log)
[bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11377103/bcr_remove_hard_restrictions.log)

**Comment by ha33y3** [comment]
> Would a couple folks mind testing this build? [BCR-1.41.r16.g897e5fb-release.zip](https://github.com/chenxiaolong/BCR/files/11376910/BCR-1.41.r16.g897e5fb-release.zip) This build reduces the amount of hacky stuff BCR does to work around this issue, but hopefully it still works.
> 
> Please report if BCR is working or not with this build and also upload the 2 BCR logs from `/data/local/tmp/` (even if BCR is working).
> 
> Thanks!
> 
> EDIT: No need to do anything fancy (eg. downgrade, clean flash, etc). Just install as normal.

I just checked it works like a magic

**Comment by phyrz91** [comment]
Works fine.

[bcr_clear_package_manager_caches.log](https://github.com/chenxiaolong/BCR/files/11377479/bcr_clear_package_manager_caches.log)
[bcr_remove_hard_restrictions.log](https://github.com/chenxiaolong/BCR/files/11377480/bcr_remove_hard_restrictions.log)

**Comment by chenxiaolong** [comment]
Thanks for testing! The fix has been merged in #323 and will be included in the next version (1.42).
